Price-based slippage calculation for increase liquidity

odcheung opened the pull request:
- Deprecated the old `increaseLiquidityQuoteByInputToken` quote function & added a new `increaseLiquidityQuoteByInputTokenUsingPriceSlippage`
- Adding `increaseLiquidityQuoteByLiquidity` quote function to generate a quote by a set liquidity value
- This new quote functions will determine `tokenMax` based on price movement, rather than the old method of setting the slippage threshold on the token estimation amounts.

yugure-orca commented on sdk/src/quotes/public/increase-liquidity-quote.ts:
```diff
@@ -59,7 +61,7 @@ export type IncreaseLiquidityQuote = IncreaseLiquidityInput & { tokenEstA: BN; t
  * @returns An IncreaseLiquidityInput object detailing the required token amounts & liquidity values to use when calling increase-liquidity-ix.
  */
-export function increaseLiquidityQuoteByInputToken(
+export function increaseLiquidityQuoteByInputToken_PriceSlippage(
```
I think SDK uses lower camel case for names.
Is there strong reason to use underscore here ?
One thing came to mind is `UsingPriceSlippage`.

yugure-orca commented on sdk/tests/sdk/whirlpools/quote/increase-liquidity-quote.test.ts:
```diff
@@ -1,10 +1,226 @@
+
+    function testVariation(params: { pTickLowerIndex: number, pTickUpperIndex: number, tickCurrentIndex: number, inputTokenAmount: BN, isTokenA: boolean, slippage: Percentage }) {
+      const { pTickLowerIndex, pTickUpperIndex, tickCurrentIndex, inputTokenAmount, isTokenA } = params;
+      const slippage = Percentage.fromFraction(1, 100);
```
Does this override `slippage` param, and always 1% slippage is used ?

odcheung replied:
removed!

yugure-orca commented on sdk/tests/sdk/whirlpools/quote/increase-liquidity-quote.test.ts:
```diff
@@ -1,10 +1,226 @@
+        tickLowerIndex: pTickLowerIndex,
+        tickUpperIndex: pTickUpperIndex,
+        tickCurrentIndex,
+        slippageTolerance: Percentage.fromFraction(1, 100),
```
`slippage` ?

yugure-orca commented on sdk/src/quotes/public/increase-liquidity-quote.ts:
```diff
@@ -109,32 +143,41 @@ export function increaseLiquidityQuoteByInputTokenWithParams(
 }
 
-/*** Private ***/
+function adjustPriceForSlippage(
```
hmm, since sqrtPrice is a square root, I would think that when this value is multiplied by 4, the actual price would move by a factor of 2. 
How about applying slippage to price instead of sqrtPrice or multiplied by the square of slippage ? 🤔

yugure-orca commented on sdk/src/quotes/public/increase-liquidity-quote.ts:
```diff
@@ -101,6 +104,37 @@ export function increaseLiquidityQuoteByInputTokenWithParams(
+    tokenMaxB: BN.max(BN.max(lowerQuote.tokenEstB, upperQuote.tokenEstB), quote.tokenEstB),
+    tokenEstA: quote.tokenEstA,
+    tokenEstB: quote.tokenEstB,
+    liquidityAmount: quote.liquidityAmount,
```
in trade context, "slippage" is used to fail transaction if the amount is beyond the threshold.
And it will work to protect users from sandwitch atack.

I think in the following case tokenEstA will be larger than required and deposit will succeed even if the price move to lower than slippage applied price.

- position range [92, 94] USDC/SOL
- current price: 93
- lowerAdjustedPrice: 93 x 99 = 92.07
- upperAdjustedPrice: 93 x 101 = 93.93
- input token: USDC
- input amount: 100 USDC

lowerQuote.liquidity will be calculated to deposit 100 USDC between [92, 92.07]. It is tight and results higher liquidity.
Higher liquidity will require more SOL distributed to 92.07 ~ 94.
The SOL amount will be larger than required SOL amount at 92.07 with quote.liquidity (smaller than lowerQuote.liquidity)

Since the Whirlpool contract requires a fixed liquidity input, I think we need to have maxA and maxB as the minimum amount of tokens that can achieve "liquidity" even if price moved between [current price - slippage, current price + slippage], WDYT?

odcheung replied:
The new function `increaseLiquidityQuoteByInputTokenUsingPriceSlippage`'s only job now is to infer liquidity from the input token amount and then call `increaseLiquidityQuoteByLiquidityWithParams`.

`increaseLiquidityQuoteByLiquidityWithParams` handles the slippage calculation as well. 

Each new quote function now has it's own set of test suite to verify tokenEst & tokenMax values.

yugure-orca reviewed (changes requested):
Added some comments 🙏

yugure-orca commented on sdk/tests/sdk/whirlpools/quote/increase-liquidity-quote-by-liq.test.ts:
```diff
@@ -0,0 +1,275 @@
+  }
+
+  if (currentTickIndex < lowerTickIndex) {
+    return getTokenAFromLiquidity(liquidity, sqrtPrice, upperSqrtPrice, true);
```
``(liquidity, lowerSqrtPrice, upperSqrtPrice)`` ?

yugure-orca commented on sdk/tests/sdk/whirlpools/quote/increase-liquidity-quote-by-liq.test.ts:
```diff
@@ -0,0 +1,275 @@
+  [Percentage.fromFraction(1, 1000), new BN(17733543)] as const,
+  [Percentage.fromFraction(1, 100), new BN(17733543)] as const,
+];
+
```
Generally same review comments for increase-liquidity-quote-by-input-token.test.ts 🙏

yugure-orca reviewed (commented):
Added some comments 🙏

odcheung commented:
@yugure-orca - Apologies. I've added the following comment to clarify what each test-case means:

```
// [---P---] = P is the current price & [] is the slippage boundary
// |-------| = Position Boundary
```
I've also cleaned up and modified the test cases to use relative ticks to the current tick rather than absolute ticks. Additional test cases with current tick variations have also been added. These will address your comments w.r.t test comments above. Let me know if I'm missing ssomething.

0% slippage test also has been removed as it doesn't apply to these test cases that tests position & slippage range overlaps.

yugure-orca reviewed (approved):
Thanks for updating.
Added some comments but I believe these are not critical comments.
